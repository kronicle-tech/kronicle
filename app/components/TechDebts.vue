<template>
  <ul v-if="techDebts && techDebts.length > 0">
    <li v-for="(techDebt, techDebtIndex) in techDebts" :key="techDebtIndex">
      <span v-if="techDebt.priority === 'very-high'">
        <b-badge variant="danger">Very High</b-badge>
      </span>
      <span v-else-if="techDebt.priority === 'high'">
        <b-badge variant="warning">High</b-badge>
      </span>
      <span v-else-if="techDebt.priority === 'medium'">
        <b-badge variant="info">Medium</b-badge>
      </span>
      <span v-else-if="techDebt.priority === 'low'">
        <b-badge variant="primary">Low</b-badge>
      </span>
      <span v-else>
        <b-badge variant="success">Missing Priority</b-badge>
      </span>
      <Markdown :markdown="techDebt.description" />
      <Markdown :markdown="techDebt.notes" />
      <Links :links="techDebt.links" />
    </li>
  </ul>
</template>

<script lang="ts">
import Vue, { PropType } from 'vue'
import { BBadge } from 'bootstrap-vue'
import { TechDebt } from '~/types/component-catalog-service'
import Links from '~/components/Links.vue'
import Markdown from '~/components/Markdown.vue'

export default Vue.extend({
  components: {
    'b-badge': BBadge,
    Links,
    Markdown,
  },
  props: {
    techDebts: {
      type: Array as PropType<TechDebt[]>,
      default: undefined,
    },
  },
})
</script>
