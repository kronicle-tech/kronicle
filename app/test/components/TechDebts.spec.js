import { mount } from '@vue/test-utils'
import TechDebts from '@/components/TechDebts.vue'

describe('TechDebts', () => {
  let propsData
  let wrapper
  const createWrapper = () => {
    wrapper = mount(TechDebts, { propsData })
  }

  beforeEach(() => {
    propsData = {}
  })

  afterEach(() => {
    wrapper.destroy()
    wrapper = null
  })

  describe('when techDebts prop is not set', () => {
    test('renders nothing', () => {
      createWrapper()
      expect(wrapper.html()).toEqual(``)
    })
  })

  describe('when techDebts prop is set to an empty array', () => {
    beforeEach(() => {
      propsData.techDebts = []
    })

    test('renders nothing', () => {
      createWrapper()
      expect(wrapper.html()).toEqual(``)
    })
  })

  describe('when techDebts prop is set to one tech debt', () => {
    beforeEach(() => {
      propsData.techDebts = [
        {
          description: 'Test Description 1',
        },
      ]
    })

    test('renders an unordered list showing the tech debt', () => {
      createWrapper()
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('when techDebts prop is set to a tech debt with very high priority', () => {
    beforeEach(() => {
      propsData.techDebts = [
        {
          description: 'Test Description 1',
          priority: 'very-high',
        },
      ]
    })

    test('renders the tech debt with a very high priority badge', () => {
      createWrapper()
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('when techDebts prop is set to a tech debt with high priority', () => {
    beforeEach(() => {
      propsData.techDebts = [
        {
          description: 'Test Description 1',
          priority: 'high',
        },
      ]
    })

    test('renders the tech debt with a high priority badge', () => {
      createWrapper()
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('when techDebts prop is set to a tech debt with medium priority', () => {
    beforeEach(() => {
      propsData.techDebts = [
        {
          description: 'Test Description 1',
          priority: 'medium',
        },
      ]
    })

    test('renders the tech debt with a medium priority badge', () => {
      createWrapper()
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('when techDebts prop is set to a tech debt with low priority', () => {
    beforeEach(() => {
      propsData.techDebts = [
        {
          description: 'Test Description 1',
          priority: 'low',
        },
      ]
    })

    test('renders the tech debt with a low priority badge', () => {
      createWrapper()
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('when techDebts prop is set to a tech debt with no priority', () => {
    beforeEach(() => {
      propsData.techDebts = [
        {
          description: 'Test Description 1',
        },
      ]
    })

    test('renders the tech debt with a missing priority badge', () => {
      createWrapper()
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('when techDebts prop is set to a tech debt with links', () => {
    beforeEach(() => {
      propsData.techDebts = [
        {
          description: 'Test Description 1',
          links: [
            {
              url: 'http://example.com/test-link-1',
            },
            {
              url: 'http://example.com/test-link-2',
            },
          ],
        },
      ]
    })

    test('renders the tech debt with the links', () => {
      createWrapper()
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('when techDebts prop is set to one tech debt with Markdown in the description', () => {
    beforeEach(() => {
      propsData.techDebts = [
        {
          description: 'Text with *bold* formatting',
        },
      ]
    })

    test('renders the tech debt with the Markdown converted to HTML', () => {
      createWrapper()
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('when techDebts prop is set to multiple tech debts', () => {
    beforeEach(() => {
      propsData.techDebts = [
        {
          description: 'Test Description 1',
        },
        {
          description: 'Test Description 2',
        },
        {
          description: 'Test Description 3',
        },
      ]
    })

    test('renders an unordered list showing the tech debts', () => {
      createWrapper()
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
})
