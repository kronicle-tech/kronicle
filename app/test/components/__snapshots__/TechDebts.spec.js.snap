// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TechDebts when techDebts prop is set to a tech debt with high priority renders the tech debt with a high priority badge 1`] = `
"<ul>
  <li><span><span class=\\"badge badge-warning\\">High</span></span>
    <div>
      <p>Test Description 1</p>
    </div>
    <div></div>
    <!---->
  </li>
</ul>"
`;

exports[`TechDebts when techDebts prop is set to a tech debt with links renders the tech debt with the links 1`] = `
"<ul>
  <li><span><span class=\\"badge badge-success\\">Missing Priority</span></span>
    <div>
      <p>Test Description 1</p>
    </div>
    <div></div>
    <ul>
      <li><a href=\\"http://example.com/test-link-1\\" target=\\"_blank\\">http://example.com/test-link-1</a> <br>
        <!---->
      </li>
      <li><a href=\\"http://example.com/test-link-2\\" target=\\"_blank\\">http://example.com/test-link-2</a> <br>
        <!---->
      </li>
    </ul>
  </li>
</ul>"
`;

exports[`TechDebts when techDebts prop is set to a tech debt with low priority renders the tech debt with a low priority badge 1`] = `
"<ul>
  <li><span><span class=\\"badge badge-primary\\">Low</span></span>
    <div>
      <p>Test Description 1</p>
    </div>
    <div></div>
    <!---->
  </li>
</ul>"
`;

exports[`TechDebts when techDebts prop is set to a tech debt with medium priority renders the tech debt with a medium priority badge 1`] = `
"<ul>
  <li><span><span class=\\"badge badge-info\\">Medium</span></span>
    <div>
      <p>Test Description 1</p>
    </div>
    <div></div>
    <!---->
  </li>
</ul>"
`;

exports[`TechDebts when techDebts prop is set to a tech debt with no priority renders the tech debt with a missing priority badge 1`] = `
"<ul>
  <li><span><span class=\\"badge badge-success\\">Missing Priority</span></span>
    <div>
      <p>Test Description 1</p>
    </div>
    <div></div>
    <!---->
  </li>
</ul>"
`;

exports[`TechDebts when techDebts prop is set to a tech debt with very high priority renders the tech debt with a very high priority badge 1`] = `
"<ul>
  <li><span><span class=\\"badge badge-danger\\">Very High</span></span>
    <div>
      <p>Test Description 1</p>
    </div>
    <div></div>
    <!---->
  </li>
</ul>"
`;

exports[`TechDebts when techDebts prop is set to multiple tech debts renders an unordered list showing the tech debts 1`] = `
"<ul>
  <li><span><span class=\\"badge badge-success\\">Missing Priority</span></span>
    <div>
      <p>Test Description 1</p>
    </div>
    <div></div>
    <!---->
  </li>
  <li><span><span class=\\"badge badge-success\\">Missing Priority</span></span>
    <div>
      <p>Test Description 2</p>
    </div>
    <div></div>
    <!---->
  </li>
  <li><span><span class=\\"badge badge-success\\">Missing Priority</span></span>
    <div>
      <p>Test Description 3</p>
    </div>
    <div></div>
    <!---->
  </li>
</ul>"
`;

exports[`TechDebts when techDebts prop is set to one tech debt renders an unordered list showing the tech debt 1`] = `
"<ul>
  <li><span><span class=\\"badge badge-success\\">Missing Priority</span></span>
    <div>
      <p>Test Description 1</p>
    </div>
    <div></div>
    <!---->
  </li>
</ul>"
`;

exports[`TechDebts when techDebts prop is set to one tech debt with Markdown in the description renders the tech debt with the Markdown converted to HTML 1`] = `
"<ul>
  <li><span><span class=\\"badge badge-success\\">Missing Priority</span></span>
    <div>
      <p>Text with <em>bold</em> formatting</p>
    </div>
    <div></div>
    <!---->
  </li>
</ul>"
`;
